overlord/snapshotstate/backend/backend.go

```diff
@@ -378,7 +387,54 @@ func Save(ctx context.Context, id uint64, si *snap.Info, cfg map[string]interfac
 
 var isTesting = snapdenv.Testing()
 
-func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer, username string, entry, dir string) error {
+type addDirToZipOptions struct {
+	ExcludePaths []string `yaml:"exclude"`
+}
+
+func readSnapshotYaml(si *snap.Info, opts *addDirToZipOptions) error {
+	file, err := osOpen(filepath.Join(si.MountDir(), "meta", "snapshots.yaml"))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		} else {
+			return nil
+		}
+	}
+	defer file.Close()
+
+	if err := yaml.NewDecoder(file).Decode(opts); err != nil {
+		return fmt.Errorf("cannot read snapshot manifest: %v", err)
+	}
+
+	// Validate the exclude list; note that this is an *exclusion* list, so
+	// even if the manifest specified paths starting with ../ this would not
+	// cause tar to navigate into those directories and pose a security risk.
+	// Still, let's have a minimal validation on them being sensible.
+	validFirstComponents := []string{
+		"$SNAP_DATA", "$SNAP_COMMON", "$SNAP_USER_DATA", "$SNAP_USER_COMMON",
+	}
+	for _, excludePath := range opts.ExcludePaths {
+		firstSlash := strings.IndexRune(excludePath, '/')
+		var firstComponent string
+		if firstSlash < 0 {
+			firstComponent = excludePath
+		} else {
+			firstComponent = excludePath[:firstSlash]
+		}
+		if !strutil.ListContains(validFirstComponents, firstComponent) {
+			return fmt.Errorf("snapshot exclude path must start with one of %q (got: %q)", validFirstComponents, excludePath)
+		}
+
+		cleanPath := filepath.Clean(excludePath)
+		if cleanPath != excludePath {
+			return fmt.Errorf("snapshot exclude path not clean: %q", excludePath)
+		}
+	}
+
```

Comment:
I think we need to be a bit more restrictive here when we validate. Right now we will accept patterns like: `- $SNAP_DATA/foo/**/bar` but AFAICT tar does not support globstar and this will work work like people will expect  (i.e. exclude anything under `$SNAP_DATA/foo/` in any subdir that is called `bar`. So my suggestion is to detect globar and simply error for now until we support them.

Reply:
I did not address this, I think we should though